[sienna.o] 방탈출 예약 관리 1, 2단계

binimini opened the pull request:
안녕하세요 맷! 이번 미션부터는 스프링을 사용하게 되네요 🙃

이번 미션의 요구사항은 방탈출 예약을 생성/조회/삭제 해주는 프로그램이 기존에 콘솔로 구현되어있고,
해당 콘솔 프로그램을 동작하게 유지하면서 웹(API)으로도 작동할 수 있도록 구현하기였습니다!

웹과 콘솔은 각각 데이터베이스를 다른 방식으로 사용하는데 로직 (방탈출 예약 생성/조회/삭제) 자체는 동일합니다.
- 콘솔 데이터베이스 : `Jdbc` 
- 웹 데이터베이스 : `Jdbc Template`

### 구현 내용
- `ReservationRepository` : 예약 model 레포지토리 인터페이스 
  - `JdbcReservationRepository` : 콘솔에서 사용하는 레포지토리 구현체
  - `JdbcTemplateReservationRepository` : 웹에서 사용하는 레포지토리 구현체
- 웹은 Application, Controller, Service, Repository로 계층을 나눠 구현했습니다!
  - 예외 발생시 응답 반환을 위한 ControllerAdvice를 추가했습니다.
- 콘솔은 입출력이나 처리 흐름을 바로 파악할 수 있도록 데이터베이스 외에는 별도 계층으로 구분하지 않았습니다.

Jdbc 사용이 처음이라 아직 공부해야할 게 많은 것 같습니다..ㅎㅎ
오늘도 리뷰 감사합니다 🙇‍♂️

binimini committed:
docs: 기능 목록 작성

binimini committed:
feat: 예약 생성 기능 구현

binimini committed:
feat: 예약 조회 및 삭제 기능 구현

binimini committed:
feat: 예약 예외 처리 구현

binimini committed:
feat: 콘솔 데이터베이스 적용

binimini committed:
feat: 웹 데이터베이스 적용

binimini committed:
refactor: 네이밍 통일 및 필요없는 코드 삭제

binimini committed:
refactor: 웹 dto 패키지 분리

binimini committed:
refactor: 콘솔 함수로 분리

- 입출력 함수로 분리
- 예약 생성 / 조회 / 삭제 함수로 분리
- 같은 시간대의 예약 존재하는 경우 예외 처리
- 해당 번호의 예약이 존재하지 않는 경우 예외 처리

binimini committed:
test: 테스트 개선

- 중복 제거
- 테스트를 위한 상수들 변수로 분리

binimini committed:
refactor: ReservationRepository 인터페이스 추가 및 중복 코드 분리

- ReservationConverter : Jdbc에서 Reservation 필드를 전달 / 반환 코드의 중복 제거
- ReservationRepository : 콘솔과 웹에서 모두 사용하는 레포 함수 정의
- JdbcReservationRepository, JdbcTemplateReservationRepository : ReservationRepository의 구현체

binimini committed:
refactor: 레포지토리 타입 인터페이스로 교체

- 구현체를 직접 명시 -> 인터페이스 명시 후 구현체 주입

binimini committed:
refactor: try/catch 한 번에 처리하도록 수정

binimini committed:
refactor: mapper 변수로 분리

binimini committed:
feat: 웹 ControllerAdvice 구현

- Controller단에서 발생하는 예외 잡아서 ResponseEntity 반환

matt-jylee commented on src/main/java/nextstep/console/RoomEscapeApplication.java:
```diff
@@ -0,0 +1,116 @@
+        String params = input.split(" ")[1];
+        Long id = Long.parseLong(params.split(",")[0]);
+
+        Optional<Reservation> result = reservationRepository.findById(id);
+        if (result.isEmpty()) {
+            throw new ReservationNotFoundException(id);
+        }
+        return result.get();
```
중요한것은 아닙니다.
자바의 Optional에는 여기 표현을 간략하게 줄이는데 도움을 주는 메서드를 제공합니다!
확인해 볼까요?!

binimini replied:
오 `orElseThrow()` 사용할 수 있는 부분이군요! 적용해보겠습니다~

matt-jylee commented on src/main/java/nextstep/model/Theme.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.Objects;
+
+public class Theme {
```
이것도 중요하지 않지만, 사람을 위해 가끔 toString을 override 하기도 합니다. ㅎㅎ

binimini replied:
3단계로 테마 확장을 구현할 저를 위해서 추가해놔야겠네요ㅎㅎ

matt-jylee commented on src/main/java/nextstep/repository/ReservationConverter.java:
```diff
@@ -0,0 +1,29 @@
+import java.time.LocalDate;
+import java.time.LocalTime;
+
+public class ReservationConverter {
```
이 유틸 클래스는 분리된 이유가 무엇일지 궁금합니다.

binimini replied:
`ReservationRepository` 인터페이스를 구현하는 구현체들 간에 중복된 코드를 제거하기 위한 클래스입니다.
하지만 단순히 중복된 코드의 분리를 위한 클래스라서 해당 클래스를 만드는 게 좋은 것인지 고민이 되긴합니다.
구현체 간의 중복을 해결하는 더 좋은 방법이 있나요? 
아니면 코드가 중복되더라도 명확한 이해를 위해 중복을 감수하는 게 나을까요?

matt-jylee replied:
중복이 허용되는 경우도 물론 있습니다만, 우리는 아직 그 중복을 허용할 단계까지 가지 못했습니다.
적어도 한 프로젝트 내에서는 의도가 같은데 중복을 허용하면 고통을 받을 확률이 매우 높아 집니다.

덧붙여, 설명해주신 의도로 유틸리티를 작성하셨다면 이름도 의도에 맞게 변해야 할것 같습니다.
영어를 못하는 저에게 Converter가 get과 set을 할 수 있다는 것이 조금 어색하게 느껴집니다.

binimini replied:
좀 더 의미를 살릴 수 있는 네이밍으로 변경해보겠습니다!

matt-jylee commented on src/main/java/nextstep/web/RoomEscapeController.java:
```diff
@@ -0,0 +1,36 @@
+import java.net.URI;
+
+@RestController
+public class RoomEscapeController {
```
상태에 맞는 응답 코드를 잘 적용하셨네요! 좋습니다!

matt-jylee commented on src/main/java/nextstep/web/RoomEscapeControllerAdvice.java:
```diff
@@ -0,0 +1,15 @@
+public class RoomEscapeControllerAdvice {
+    @ExceptionHandler(RoomEscapeException.class)
+    public ResponseEntity<String> handleRoomEscapeException(RoomEscapeException e) {
+        return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
```
예외 처리도 이제 중요해지는데, 모든 RoomEscapeException이 BAD_REQUEST로 응답되는게 의도된 것인지 궁금합니다.

binimini replied:
예외 클래스들을 단순히 메세지만 가지게 하다보니 응답 코드를 통일하게 되었는데,
예외 클래스에 각 상태코드와 메세지를 가지게 해서 상세한 응답을 줄 수 있는 게 더 좋은 것 같습니다! 수정해보겠습니당 😀

matt-jylee commented on src/main/java/nextstep/console/RoomEscapeApplication.java:
```diff
@@ -0,0 +1,116 @@
+        System.out.println("예약이 취소되었습니다.");
+    }
+
+    public static void printReservation(Reservation reservation) {
+        System.out.println("예약 번호: " + reservation.getId());
+        System.out.println("예약 날짜: " + reservation.getDate());
+        System.out.println("예약 시간: " + reservation.getTime());
+        System.out.println("예약자 이름: " + reservation.getName());
+    }
+
+    public static void printReservationTheme(Theme theme) {
+        System.out.println("예약 테마 이름: " + theme.getName());
+        System.out.println("예약 테마 설명: " + theme.getDesc());
+        System.out.println("예약 테마 가격: " + theme.getPrice());
+    }
```
printReservation / printReservationTheme 은 왜 나눠져 있고 어떤 경우에 따로 호출되는 걸까요?

binimini replied:
예약 생성의 경우에는 `Reservation 정보`만 출력하고,
예약 조회의 경우에는 `Reservation 정보`와 `Reservation 내부의 Theme 정보`까지 출력합니다.
각각을 함수로 만들면 중복된 코드를 작성해야할 것 같아서 Reservation과 Theme 출력을 분리해서 작성하게 되었습니다!

matt-jylee commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE IF NOT EXISTS RESERVATION
```
실무에서 99퍼센트는 id가 뭔지 theme_name 이 뭔지 유추를 하지만
모두를 위해 COMMENT 를 추가해서 뭘 의미하는 지 한글로 조금 더 설명해줍니다.

binimini replied:
오 sql 파일을 공유해야할 상황이 없어서 생각해보지 못한 부분인 것 같아요! 좋은 정보 감사합니다 😊

matt-jylee replied:
오! 의미를 한 결 더 쉽게 이해할 수 있게 되었네요!

matt-jylee reviewed (changes requested):
TMI 인터페이스 등 구현에 관한 이야기입니다.
내가 ABCInterface 를 구현한다면 관례적으로 구현 클래스의 이름에 "Impl"을 접미사로 붙이기도 합니다.
하지만 반드시 따라야 하는 규칙은 아닙니다.

interface AbcInterface {}
class AbcInterfaceImpl implements AbcInterface {}

binimini committed:
refactor: Optional 메서드 적용

- orElseThrow() 적용

binimini committed:
feat: Theme toString 구현

binimini committed:
refactor: 인터페이스 접근제한자 명시 삭제

- 인터페이스 기존 접근제한자 : public

binimini committed:
refactor: 예외별 상태코드 추가

- ReservationDuplicateException : 사용자가 자원의 충돌 해결해야함 : 409 Conflict
- ReservationNotFoundException : 해당 자원 찾을 수 없음 : 404 NotFound

binimini committed:
refactor: SQL 파일 코멘트 추가

- 컬럼에 대한 설명 추가

binimini committed:
refactor: ReservationService 공통으로 사용하도록 수정

- RoomEscapeService -> ReservationService 네이밍 수정
- 콘솔의 어플리케이션에서 파싱 코드 함수로 분리해 Service 및 DTO 공통 사용하도록 수정
- 서비스 패키지 이동으로 인해 컴포넌트 스캔 위한 웹 어플리케이션의 어노테이션 추가

binimini committed:
refactor: Controller 네이밍 수정

- RoomEscapeController -> ReservationController 네이밍 수정

binimini commented:
맷이 말씀해주신 피드백 반영했습니다!

또 콘솔 어플리케이션과 웹 어플리케이션이 레포지토리를 사용하는 로직(서비스단 로직)이 동일한데
콘솔 어플리케이션이 중복해서 작성하는 것 같아 기존의 웹 어플리케이션의 서비스를 공통으로 사용하도록 수정했습니다!
### 수정사항
- 어플리케이션 수정
  - 콘솔 어플리케이션의 파싱 코드 함수로 분리
  - 웹 어플리케이션의 서비스 -> 웹·콘솔 공통 서비스로 수정
  - `RoomEscapeController` -> `ReservationController`, `RoomEscapeService` -> `ReservationService` 네이밍 수정
- Optional 메서드 적용
- `theme.toString()` 구현
- 예외별 상태코드 응답하도록 수정
- SQL 코멘트 작성

matt-jylee reviewed (changes requested):
유틸리티 클래스를 한번 더 고민해보면 좋겠습니다.

binimini committed:
refactor: 중복 코드 제거를 위한 유틸 클래스 네이밍 및 디렉토리 수정

- ReservationConverter -> JdbcRemoveDuplicateUtils
- ReservationConverter.get() -> JdbcRemoveDuplicateUtils.getReservationFromResultSet()
- ReservationConverter.set() -> JdbcRemoveDuplicateUtils.setReservationToStatement()
- util 디렉토리로 이동
- 세부적인 함수명으로 변경

matt-jylee reviewed (approved):
수고하셨습니다!

3단계에서 뵙겠습니다!
